pkg/services/sqldb/provision.go

```diff
@@ -95,16 +176,96 @@ func (s *serviceManager) preProvision(
 	return dt, nil
 }
 
-func buildARMTemplateParameters(
+func (v *vmOnlyManager) preProvision(
+	_ context.Context,
+	instance service.Instance,
+	_ service.Plan,
+	_ service.Instance, //reference instance
+) (service.InstanceDetails, error) {
+	dt, ok := instance.Details.(*mssqlVMOnlyInstanceDetails)
+	if !ok {
+		return nil, errors.New(
+			"error casting instance.Details as *mssqlVMOnlyInstanceDetails",
+		)
+	}
+	dt.ARMDeploymentName = uuid.NewV4().String()
+	dt.ServerName = uuid.NewV4().String()
+	dt.AdministratorLogin = generate.NewIdentifier()
+	dt.AdministratorLoginPassword = generate.NewPassword()
+	return dt, nil
+}
+
+func (d *dbOnlyManager) preProvision(
+	_ context.Context,
+	instance service.Instance,
+	_ service.Plan,
+	referenceInstance service.Instance, //reference instance
+) (service.InstanceDetails, error) {
+	dt, ok := instance.Details.(*mssqlInstanceDetails)
+	if !ok {
+		return nil, errors.New(
+			"error casting instance.Details as *mssqlInstanceDetails",
+		)
+	}
+	//Assume refererence instance is a vm only instance?
+	rdt, ok := referenceInstance.Details.(*mssqlVMOnlyInstanceDetails)
+	if !ok {
+		return nil, errors.New(
+			"error casting referenceInstance.Details as " +
+				"*mssqlVMOnlyInstanceDetails",
+		)
+	}
+
+	azureConfig, err := azure.GetConfig()
+	if err != nil {
+		return nil, err
+	}
+	azureEnvironment, err := az.EnvironmentFromName(azureConfig.Environment)
+	if err != nil {
+		return nil, err
+	}
+
+	dt.DatabaseName = generate.NewIdentifier()
+
+	//Build the instance details with the reference instance details
+	dt.ServerName = rdt.ServerName
+	dt.AdministratorLogin = rdt.AdministratorLogin
+	dt.AdministratorLoginPassword = rdt.AdministratorLoginPassword
```

Comment:
I'm thinking a db-only instance doesn't need to keep track of the three properties above itself, right?

Reply:
Probably not server name, but we don’t pass refer node instance to bind/unbind and we need to have the admin login/password for creating and removing the binding no?

Reply:
You're right. But I think right thing to do then is probably to start passing ref instance when we bind.

Reply:
You know what... what would be even better is if instances had a ref instance _field_. So we never have to pass ref instances around at all. If you have an instance, you'll already have the _means_ to obtain its ref instance (if it has one).

I'll tackle that.

In the meantime, let's not let this issue block us from merging this.